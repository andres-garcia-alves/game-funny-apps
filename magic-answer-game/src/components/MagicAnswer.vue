<template>
    <div>
        <img v-if="image" :src="image" alt="bg">
        <div class="bg-dark"></div>
        <div class="magic-answer-container">
            <input type="text" placeholder="Estás indeciso? Preguntame lo que quieras..." v-model=question>
            <p>Para terminar la pregunta usa un signo de interrogación '?'</p>

            <div v-if="questionText">
                <h3 class="question">{{ questionText }}</h3>
                <h1 class="answer">{{ answer }}</h1> 
            </div>
        </div>
    </div>
</template>

<script>
export default {
    name: 'MagicAnswer',
    data() {
        return {
            question: null,
            questionText: null,
            answer: null,
            image: null
        }        
    },
    watch: {
        async question(value, oldValue) {
            console.log({ value })
            if (value.includes('?') == false) { return }

            await this.getAnswer();

            this.questionText = this.question
            this.question = ''
        }
    },
    expose: ['getAnswer', 'answer', 'image'],
    methods: {
        async getAnswer() {
            try {
                this.answer = 'pensando...'

                const response = await fetch('https://yesno.wtf/api')
                const { answer, image } = await response.json()

                this.image = image

                if (answer === 'yes')           { this.answer = 'Si ;-)' }
                else if (answer === 'no')       { this.answer = 'No!' }
                else if (answer === 'maybe')    { this.answer = 'Tal vez' }
                else                            { this.answer = 'Ni idea' }
            }
            catch (error) {
                console.log('Error', error)
                this.answer = 'Error en el llamado a la API'
                this.image = null
            }
        }
    }
}
</script>

<style scoped>
img, .bg-dark {
    height: 100vh;
    left: 0px;
    max-height: 100%;
    max-width: 100%;
    position: fixed;
    top: 0px;
    width: 100vw;
}
p {
    color: white;
    font-size: 12px;
    margin-top: 0px;
}
input:focus {
    outline: none;
}
.bg-dark {
    background-color: rgba(0, 0, 0, 0.4);
}
.magic-answer-container {
    position: relative;
    z-index: 99;
}
.magic-answer-container input {
    width: 400px;
    padding: 10px 15px;
    border-radius: 5px;
    border: none;
}
.magic-answer-container p {
    margin-top: 25px;
}
.magic-answer-container .question {
    color: white;
    margin-top: 120px;
}
.magic-answer-container .answer {
    color: white;
    margin-top: 50px;
}
</style>
