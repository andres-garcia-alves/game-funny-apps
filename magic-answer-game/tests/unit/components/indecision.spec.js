import { shallowMount } from '@vue/test-utils'
import MagicAnswer from '@/components/MagicAnswer.vue'

describe("MagicAnswer.vue", () => {
    let wrapper, consoleSpy;

    global.fetch = jest.fn( () => Promise.resolve({
        json: () => Promise.resolve({
            answer: 'yes',
            forced: false,
            image: 'https://yesno.wtf/assets/yes/2.gif'
        })
    }))

    beforeEach(() => {
        wrapper = shallowMount(MagicAnswer)

        consoleSpy = jest.spyOn(console, 'log') // console.log()
        jest.clearAllMocks()
    })

    test("debe de hacer match con el snapshot", () => {
        expect(wrapper.html()).toMatchSnapshot()
    })

    test("no debe de cambiar nada hasta que se escriba el caracter '?'", async () => {
        const getAnswerSpy = jest.spyOn(wrapper.vm, 'getAnswer')            // necesita "expose: ['getAnswer']" en el componente
        // const getAnswerSpy = jest.spyOn(MagicAnswer.methods, 'getAnswer') // equivale a MagicAnswer.getAnswer()

        const input = wrapper.find('input')
        await input.setValue('Hola mundo')

        expect(consoleSpy).toHaveBeenCalledTimes(1)
        expect(getAnswerSpy).toHaveBeenCalledTimes(0)
    })

    test("debe de disparar un getAnswer() cuendo se escriba el caracter '?'", async () => {
        const getAnswerSpy = jest.spyOn(wrapper.vm, 'getAnswer')            // necesita "expose: ['getAnswer']" en el componente
        // const getAnswerSpy = jest.spyOn(MagicAnswer.methods, 'getAnswer') // equivale a MagicAnswer.getAnswer()
        
        const input = wrapper.find('input')
        await input.setValue('Hola mundo?')

        expect(consoleSpy).toHaveBeenCalledTimes(1)
        expect(getAnswerSpy).toHaveBeenCalledTimes(1)
    })

    test("debe procesarse la respuesta de la API externa", async () => {

        // arrange & act
        await wrapper.vm.getAnswer()        
        const image = wrapper.find('img')

        // asserts
        expect(image.exists()).toBe(true)
        expect(wrapper.vm.answer).toBe('Si ;-)')
        expect(wrapper.vm.image).toBe('https://yesno.wtf/assets/yes/2.gif')
    })

    test("debe validarse un fallo en la respuesta de la API externa", async () => {

        // arrange & act
        fetch.mockImplementationOnce( () => Promise.reject('API is down'))

        await wrapper.vm.getAnswer()   
        const image = wrapper.find('img')

        // asserts
        expect(image.exists()).toBe(false)
        expect(wrapper.vm.answer).toBe('Error en el llamado a la API')
    })
})
