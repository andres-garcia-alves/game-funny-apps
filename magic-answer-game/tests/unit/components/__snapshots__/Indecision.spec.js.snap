// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MagicAnswer.vue debe de hacer match con el snapshot 1`] = `
<div>
  <!--v-if-->
  <div class="bg-dark"></div>
  <div class="magic-answer-container"><input type="text" placeholder="Estás indeciso? Preguntame lo que quieras...">
    <p>Para terminar la pregunta usa un signo de interrogación '?'</p>
    <!--v-if-->
  </div>
</div>
`;
